ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/query/TransactionLogBloomCacher.java

```diff
@@ -128,27 +132,63 @@ private void fillCacheFile(
     }
   }
 
-  public void cacheLogsBloomForBlockHeader(
+  void cacheLogsBloomForBlockHeader(
       final BlockHeader blockHeader,
       final Optional<File> reusedCacheFile,
       final boolean ensureChecks) {
     try {
+      if (cachingStatus.cachingCount.incrementAndGet() != 1) {
+        return;
+      }
       final long blockNumber = blockHeader.getNumber();
       LOG.debug("Caching logs bloom for block {}.", "0x" + Long.toHexString(blockNumber));
       if (ensureChecks) {
         ensurePreviousSegmentsArePresent(blockNumber);
       }
       final File cacheFile = reusedCacheFile.orElse(calculateCacheFileName(blockNumber, cacheDir));
-      if (!cacheFile.exists()) {
-        Files.createFile(cacheFile.toPath());
+      if (cacheFile.exists()) {
+        cacheSingleBlock(blockHeader, cacheFile);
+      } else {
+        scheduler.scheduleComputationTask(this::populateLatestSegment);
+      }
+    } catch (final IOException e) {
+      LOG.error("Unhandled caching exception.", e);
+    } finally {
+      cachingStatus.cachingCount.decrementAndGet();
+    }
+  }
+
+  private void cacheSingleBlock(final BlockHeader blockHeader, final File cacheFile)
+      throws IOException {
+    try (final RandomAccessFile writer = new RandomAccessFile(cacheFile, "rw")) {
+      final long offset = (blockHeader.getNumber() % BLOCKS_PER_BLOOM_CACHE) * BLOOM_BITS_LENGTH;
+      writer.seek(offset);
+      writer.write(ensureBloomBitsAreCorrectLength(blockHeader.getLogsBloom().toArray()));
+    }
+  }
+
+  private boolean populateLatestSegment() {
+    try {
+      long blockNumber = blockchain.getChainHeadBlockNumber();
+      final File currentFile = calculateCacheFileName(CURRENT, cacheDir);
+      final long segmentNumber = blockNumber / BLOCKS_PER_BLOOM_CACHE;
+      try (final OutputStream out = new FileOutputStream(currentFile)) {
+        fillCacheFile(segmentNumber * BLOCKS_PER_BLOOM_CACHE, blockNumber, out);
       }
-      try (RandomAccessFile writer = new RandomAccessFile(cacheFile, "rw")) {
-        final long offset = (blockNumber / BLOCKS_PER_BLOOM_CACHE) * BLOOM_BITS_LENGTH;
-        writer.seek(offset);
-        writer.write(ensureBloomBitsAreCorrectLength(blockHeader.getLogsBloom().toArray()));
+      while (blockNumber <= blockchain.getChainHeadBlockNumber()
+          && (blockNumber % BLOCKS_PER_BLOOM_CACHE != 0)) {
+        cacheSingleBlock(blockchain.getBlockHeader(blockNumber).orElseThrow(), currentFile);
+        blockNumber++;
```

Comment:
This `while` loop near the chain head on `EthScheduler` seems to be approximating a listener. The whole logs caching endeavor reminds me of pruning in that way. Where there's something that needs to be done in bulk (marking the whole state trie and one block <-> doing all the caching for previous segments) and one that is done with a listener (marking new nodes from new blocks with a listener on the state trie <-> random access writing the log bloom for recent blocks). Could a similar approach be useful here?

Reply:
This is catching up for the case where more blocks could have shown up in the interim while caching the current segment.  We could do it with a listener but I am concerned that the readability would further suffer, since we would need to put the newly arrived blocks to the side while we finish up caching the segment and then while we pop off a listened object yet another one could show up.  This class already is a tad bit hard to read and making a `do this while we are behind` keeps that logic isolated to one method.